server/command.js

```diff
@@ -0,0 +1,243 @@
+  var dobjstr = '';
+  var dobj = null;
+  var iobjstr = '';
+  var iobj = null;
+  var prepstr = '';
+  var m = argstr.match($.utils.command.prepositionsRegExp);
+  if (m) {
+    prepstr = m[1];
+    var i = argstr.indexOf(prepstr);
+    if (i === -1) {
+      throw RangeError('Impossible preposition: ' + prepstr);
+    }
+    dobjstr = argstr.substring(0, i - 1);
+    iobjstr = argstr.substring(i + prepstr.length + 1);
+    prepstr = prepstr.replace(/ +/g, ' ');
+  } else {
+    dobjstr = argstr;
+  }
+  if (dobjstr) {
+    var m = $.utils.match(dobjstr, user);
+    if (!m.length) {
+      dobj = undefined;
+    } else if (m.length === 1) {
+      dobj = dobj[0];
+    } else {
+      dobj = ReferenceError(dobj.length + ' matches.');
+    }
+  }
+  if (iobjstr) {
+    var m = $.utils.match(iobjstr, user);
+    if (!m.length) {
+      iobj = undefined;
+    } else if (m.length === 1) {
+      iobj = iobj[0];
+    } else {
+      iobj = ReferenceError(iobj.length + ' matches.');
+    }
+  }
+  return {
+    verb: verb,
+    argstr: argstr,
+    args: args,
+    dobjstr: dobjstr,
+    dobj: dobj,
+    prepstr: prepstr,
+    iobjstr: iobjstr,
+    iobj: iobj
+  };
+};
+
+$.utils.command.execute = function(command, user) {
+  // Parse and execute a user's command.
+  // Return true if successful, false if not.
+  var spec = $.utils.command.parse(command, user);
+  if (!spec) {
+    // TODO: Support trapping of empty commands.
+    return true;
+  }
+  // Collect all objects which could host the verb.
+  var hosts = [user, user.location, spec.dobj, spec.iobj];
+  for (var i = 0; i < hosts.length; i++) {
+    var host = hosts[i];
+    if (!host) {
+      continue;
+    }
+    // Check every verb on each object for a match.
+    var func, verb, dobj, prepstr, iobj;
+    for (var prop in host) {
+      func = host[prop];
+      if (typeof func === 'function' && (prepstr = func.prepstr) &&
+          (dobj = func.dobj) && (iobj = func.iobj && (verb = func.verb))) {
+        // This is a verb.
+        if (prepstr === 'any' || prepstr == spec.prepstr ||
+            (prepstr == 'none' && !prepstr)) {
+          if (dobj === 'any' || (dobj === 'this' && spec.dobj === host) ||
+              (dobj === 'none' && !dobj)) {
+            if (iobj === 'any' || (iobj === 'this' && spec.iobj === host) ||
+                (iobj === 'none' && !iobj)) {
+              if (spec.verb.search(verb) === 0) {
+                host[prop](spec);
+                return true;
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+  return false;
+};
+
+$.utils.match = function(str, context) {
+  // Attempt to match a string with the environmental context.
+  // Context can be anything (including null) but is normally a place or a user.
+  // Returns an array of all matching objects.
+  str = str.trim();
+  // First, check for matches against universally accessible things.
+  var m = $(str);
```

Comment:
$() has been downgraded so that it can only do property lookups.  We can look at restoring function calls at a later time.